File: packages/client/src/ecs/Aspect.ts
import { Component } from './Component'
import { Entity } from './Entity'

const flags = (components: Component[]): number => {
  let flags = 0

  components.forEach(cmp => {
    flags = bit.bxor(flags, cmp._flag)
  })

  return flags
}

export class Aspect {
  all: number
  none: number
  one: number

  constructor(
    all: Component[] = [],
    none: Component[] = [],
    one: Component[] = []
  ) {
    this.all = flags(all)
    this.none = flags(none)
    this.one = flags(one)
  }

  check = (entity: Entity): boolean => {
    const all = bit.bor(this.all, entity.components) === entity.components
    const none = bit.band(this.none, entity.components) === 0
    const one = bit.band(this.one, entity.components) !== 0

    return (one || all) && none
  }
}

export class AlwaysAspect extends Aspect {
  check = (_: any): boolean => {
    return true
  }
}

export class NeverAspect extends Aspect {
  check = (_: any): boolean => {
    return false
  }
}

File: packages/client2/lib/ecs/aspect.dart
import './entity.dart';

int flags(List<int> components) {
  return components.fold(0, (acc, flag) {
    return acc ^ flag;
  });
}

class Aspect {
  int all = 0;
  int none = 0;
  int one = 0;

  Aspect({
    List<int> all = const [],
    List<int> none = const [],
    List<int> one = const [],
  }) {
    this.all = flags(all);
    this.none = flags(none);
    this.one = flags(one);
  }

  bool check(Entity entity) {
    bool all = (this.all | entity.components) == entity.components;
    bool none = (this.none & entity.components) == 0;
    bool one = (this.one & entity.components) != 0;

    return (one || all) && none;
  }
}

class AlwaysAspect extends Aspect {
  @override
  bool check(_) => true;
}

class NeverAspect extends Aspect {
  @override
  bool check(_) => false;
}
